_pages/automatic_differentiation.md

```diff
@@ -0,0 +1,153 @@
+lengthy expressions for large computations. It is limited to closed-form
+expressions; that is, it cannot process the control flow.
+
+- **Automatic Differentiation (AD)**: Automatic Differentiation is a highly
+  efficient technique that computes derivatives of mathematical functions by
+  applying differentiation rules to every arithmetic operation in the code.
+  Automatic Differentiation can be used in two modes: 
+
+    - Forward Mode: calculates derivatives with respect to a single variable, and 
+    
+    - Reverse Mode: calculates gradients with respect to all inputs
+    simultaneously.
+
+### Automatic Differentiation in C++
+
+Automated Differentiation implementations are based on Operator Overloading or
+Source Code Transformation. C++ allows operator overloading, making it
+possible to implement Automatic Differentiation. The derivative of a function
+can be evaluated at the same time as the function itself. Automatic
+Differentiation exploits the fact that every computer calculation consists of
+elementary mathematical operations and functions, and by applying the chain
+rule recurrently, partial derivatives of arbitrary order can be computed
+accurately. Following are some of its highlights:
+
+- Automatic Differentiation can calculate derivatives without any additional
+  precision loss. 
+
+- It is not confined to closed-form expressions. 
+
+- It can take derivatives of algorithms involving conditionals, loops, and
+  recursion. 
+
+- It works without generating inefficiently long expressions. 
+
+### Automatic Differentiation Implementation with Clad - a Clang Plugin
+
+Implementing Automatic Differentiation from the ground up can be challenging.
+However, several C++ libraries and tools are available to simplify the
+process. The Compiler Research Group has been working on [Clad], a C++ library
+that enables Automatic Differentiation using the LLVM compiler infrastructure.
+It is implemented as a plugin for the Clang compiler. 
+
+[Clad] operates on Clang AST (Abstract Syntax Tree) and is capable of
+performing C++ Source Code Transformation. When Clad is given the C++ source
+code of a mathematical function, it can automatically generate C++ code for
+the computing derivatives of that function. Clad has comprehensive coverage of
+the latest C++ features and a well-rounded fallback and recovery system in
+place.
+
+**Clad's Key Features**:
+
+- Support for both, Forward Mode and Reverse Mode Automatic Differentiation.
+
+- Support for differentiation of the built-in C input arrays, built-in C/C++
+  scalar types, functions with an arbitrary number of inputs, and functions
+  that only return a single value.
+
+- Support for loops and conditionals.
+
+- Support for generation of single derivatives, gradients, Hessians, and
+  Jacobians.
+
+- Integration with CUDA for GPU programming.
+
+- Integration with Cling and ROOT for high-energy physics data analysis.
+
+### Basics of using Clad
+
+Clad provides five API functions:
```

Comment:
lets leave the Clad details on the Clad project page rather than here..

Reply:
removed the Basics of using Clad